pkg/sql/sem/builtins/jobs_generator_builtin.go

```diff
@@ -0,0 +1,172 @@
+// have access to.
+func NewJobsGenerator(
+	ctx context.Context,
+	txn *kv.Txn,
+	executor sqlutil.InternalExecutor,
+	planner eval.Planner,
+	user username.SQLUsername,
+) (eval.ValueGenerator, error) {
+	userIsAdmin, err := planner.UserHasAdminRole(ctx, user)
+	if err != nil {
+		return &jobsGenerator{}, err
+	}
+	userHasControlJobRoleOption, err := planner.HasRoleOption(ctx, roleoption.CONTROLJOB)
+	if err != nil {
+		return &jobsGenerator{}, err
+	}
+	return &jobsGenerator{
+		ctx:                         ctx,
+		txn:                         txn,
+		executor:                    executor,
+		user:                        user,
+		userIsAdmin:                 userIsAdmin,
+		userHasControlJobRoleOption: userHasControlJobRoleOption,
+		planner:                     planner,
+	}, nil
+}
+
+var jobsIterType = types.MakeLabeledTuple(
+	[]*types.T{types.Int, types.String, types.Timestamp, types.Bytes, types.Bytes, types.String,
+		types.Int, types.Bytes, types.Int, types.Int, types.Timestamp},
+	[]string{"id", "status", "created", "payload", "progress", "created_by_type",
+		"created_by_id", "claim_session_id", "claim_instance_id", "num_runs", "last_run"},
+)
+
+// ResolvedType implements the eval.ValueGenerator interface.
+func (*jobsGenerator) ResolvedType() *types.T {
+	return jobsIterType
+}
+
+const jobsTableWithUsernameQuery = `
+SELECT 
+    *,
+    crdb_internal.pb_to_json('payload', payload) ->> 'usernameProto'
+FROM system.jobs;
+`
+
+// Start implements the eval.ValueGenerator interface.
+func (gen *jobsGenerator) Start(ctx context.Context, txn *kv.Txn) error {
+	iter, err := gen.executor.QueryIteratorEx(ctx,
+		"system-jobs-scan",
+		txn,
+		sessiondata.InternalExecutorOverride{User: username.RootUserName()},
+		jobsTableWithUsernameQuery,
+	)
+	if err != nil {
+		return err
+	}
+
+	gen.iter = iter
+	return nil
+}
+
+// Next implements the eval.ValueGenerator interface.
+//
+// This method is not idempotent.
+func (gen *jobsGenerator) Next(ctx context.Context) (bool, error) {
+	for {
+		hasNext, err := gen.iter.Next(ctx)
+		if !hasNext || err != nil {
+			return hasNext, err
+		}
+
+		currentRow := gen.iter.Cur()
+		jobOwnerDatum := currentRow[gen.iter.Cur().Len()-1]
+		jobOwnerUser := username.MakeSQLUsernameFromPreNormalizedString(strings.Trim(jobOwnerDatum.String(), "'"))
+		jobOwnerIsAdmin, err := gen.planner.UserHasAdminRole(ctx, jobOwnerUser)
+		if err != nil {
+			return false, err
+		}
+
+		if !userHasJobAccess(gen.user, gen.userIsAdmin, gen.userHasControlJobRoleOption, jobOwnerUser, jobOwnerIsAdmin) {
+			continue
+		}
+
+		gen.buf = gen.iter.Cur()[:gen.iter.Cur().Len()-1]
+		return hasNext, err
+	}
+}
+
+// userHasJobAccess determines if a user can access a job based on the
+// job owner and role options.
+//
+// The user can access the row if the meet one of the conditions:
+//  1. The user is an admin.
+//  2. The job is owned by the user.
+//  3. The user has CONTROLJOB privilege and the job is not owned by
+//     an admin.
```

Comment:
This helper function is nothing but the wrapper around 4 booleans.
And it  seems to be called from 1 place.
I think it would actually be clearer if you inlined it (along with your comment) at the call site.